### PR TITLE
PE-342: CLICommand class

### PR DESCRIPTION
**Pull request by matibat**

### CODE REVIEW

File: src/CLICommand/commonContext.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+	public async getIsPrivate(): Promise<boolean> {
+		return !!(this.password || this.driveKey || (await this.getWallet().catch(() => false)));
+	}
+
+	/**
+	 * @returns {Promise<Wallet>}
+	 * Will return a wallet instance created from the seed phrase or the walletFile
+	 */
+	public async getWallet(): Promise<Wallet> {
+		const walletFile = this.getParameterValue(WalletFileParameter);
```
Comment (fedellen) [inline src/CLICommand/commonContext.ts:45]: I'm trying to test the `get-balance` command with this setup and found that `walletFile` here console logs out as `true` rather than a string file path

Comment (matibat) [reply src/CLICommand/commonContext.ts:45]: CLICommand was assuming all commands were boolean. Now you can specify the desired type, which defaults to 'single-value'. Fixed at ecb1cbc

**Review by fedellen — APPROVED** [review]
Its working on my end now, looking good!

My only question here is: Do you think composing the commands this way will make any of the interactive parts (i.e. transaction cost confirmation / secure password input) more complex in the upcoming steps?

**Comment by matibat** [comment]
@fedellen 
> My only question here is: Do you think composing the commands this way will make any of the interactive parts (i.e. transaction cost confirmation / secure password input) more complex in the upcoming steps?

If you mean about the interactive input, I would:
- Add a flag to ParameterData to make it "promptable", and
- make the CommonContext to ask for the user input when the value of a certain promptable parameter is read.

So whichever library we choose is encapsulated behind CommonContext.

**Comment by matibat** [comment]
I'm about to 
- refactor the classes to make them testable
- add tests for each component

**Comment by matibat** [comment]
Added the tests for the newly introduced objects

File: src/CLICommand/common_context.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	private readonly walletDao: WalletDAO;
+
+	/**
+	 * @returns {CommonContext}
+	 * @param {any} options The object containing the parameterName: value mapping
+	 * @param {Arweave} arweave The arweave instance
+	 * An immutable instance of CommonContext holding the parsed values of the parameters
+	 */
+	constructor(private readonly options: any, private readonly arweave: Arweave) {
+		this.walletDao = new WalletDAO(this.arweave);
```
Comment (fedellen) [inline src/CLICommand/common_context.ts:29]: Lets hand down the walletDao from index.ts so we're using a single source. The ArDrive class will be using the same walletDao

File: src/CLICommand/cli_command.test.ts [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+		expect(stubbedProgram.command.calledOnce).to.be.true;
+		expect(stubbedProgram.action.calledOnce).to.be.true;
+	});
+
+	it('The library parses the given argv', () => {
+		CLICommand.parse(stubbedProgram);
+		expect(stubbedProgram.parse.calledOnce).to.be.true;
+	});
+
+	it("CLICommmand won't allow to declare a command after parsed", () => {
```
Comment (fedellen) [inline src/CLICommand/cli_command.test.ts:78]: Nit: triple mmm's in CLICommand here

File: src/CLICommand/cli_command.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+		commandDescriptor.action(options);
+	});
+}
+
+export class CLICommand {
+	private static _doneSettingCommands = false;
+	private static _argv?: string[]; // Custom argv vector for testing propuse
+
+	/**
+	 * @param {CommandDescriptor} commandDescription an immputable representation of a command
```
Comment (fedellen) [inline src/CLICommand/cli_command.ts:46]: spelling nit: "immutable", and below/above on _argv -- "purposes"

File: src/CLICommand/parameter.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+		return this.parameterData.type || 'single-value';
+	}
+
+	public static declare(parameter: ParameterData): void {
+		Parameter.parameters.push(parameter);
+	}
+
+	/**
+	 * @name reset
+	 * For testing pruposes only. It will just remove all parameters declaration
```
Comment (fedellen) [inline src/CLICommand/parameter.ts:45]: purposes, similar to other nits

File: src/CLICommand/parameter.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+export interface ParameterData {
+	name: ParameterName;
+	aliases: ParameterName[];
+	description: string;
+	type?: ParameterType;
+	default?: string;
+	required?: true;
+	forbiddenConjunctionParameters?: ParameterName[];
+	requireedConjunctionParameters?: ParameterName[];
```
Comment (fedellen) [inline src/CLICommand/parameter.ts:12]: spelling nit: extra E in required

**Review by fedellen — CHANGES_REQUESTED** [review]
Just minor things so far. I ran the tests sucessfully and self tested the get-balance command. I'll dig deeper later today

File: src/CLICommand/common_context.ts [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+			return await this.walletDao.generateJWKWallet(seedPhrase);
+		}
+		throw new Error('No wallet file neither seed phrase provided!');
+	}
+
+	/**
+	 * @type {string | undefined}
+	 * A getter for the drive password
+	 */
+	public get password(): string | undefined {
```
Comment (fedellen) [inline src/CLICommand/common_context.ts:62]: Do we need to make another getter here for each parameter we declare in parameter_declarations?

File: src/CLICommand/common_context.ts (hunk 2 above)
Comment (matibat) [reply src/CLICommand/common_context.ts]: Applied this at at 9ce6e5fa3a937ce4c9e489b8ba6a9bc9dee16b37

File: src/CLICommand/common_context.ts (hunk 7 above)
Comment (matibat) [reply src/CLICommand/common_context.ts]: Not really, it's pointless to have those as we have one generic `getParameterValue()`
Removed useless code: bd5eda138d074f05d03c853bb4b808a5f5f5df2d

File: src/CLICommand/cli_command.test.ts (hunk 3 above)
Comment (matibat) [reply src/CLICommand/cli_command.test.ts]: c0f61a7a2ad689c94578a24172c72cb016135646

File: src/CLICommand/cli_command.ts (hunk 4 above)
Comment (matibat) [reply src/CLICommand/cli_command.ts]: c0f61a7a2ad689c94578a24172c72cb016135646

File: src/CLICommand/parameter.ts (hunk 5 above)
Comment (matibat) [reply src/CLICommand/parameter.ts]: c0f61a7a2ad689c94578a24172c72cb016135646

File: src/CLICommand/parameter.ts (hunk 6 above)
Comment (matibat) [reply src/CLICommand/parameter.ts:12]: c0f61a7a2ad689c94578a24172c72cb016135646

File: src/CLICommand/common_context.ts [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+ * @type {CommonContext}
+ * A class representing the context of the parameters
+ */
+export class CommonContext {
+	private readonly walletDao: WalletDAO;
+
+	/**
+	 * @returns {CommonContext}
+	 * @param {any} options The object containing the parameterName: value mapping
+	 * @param {Arweave} arweave The arweave instance
```
Comment (fedellen) [inline src/CLICommand/common_context.ts:23]: Arweave is no longer a param here

File: src/CLICommand/cli_command.test.ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+	name: driveNameCommandName,
+	parameters: [DriveNameParameter],
+	action(option) {
+		/** This code here will run after argv is parsed */
+		expect(option.driveNameTest).to.equal(MY_DRIVE_NAME);
+	}
+};
+const driveNameArgv: string[] = [...baseArgv, driveNameCommandName, '--drive-name', MY_DRIVE_NAME];
+
+// const MY_DRIVE_ADDRESS = '00000000000000000000000000000000';
```
Comment (fedellen) [inline src/CLICommand/cli_command.test.ts:24]: Lets drop the commented code here, unless you're planning on extending the test with this.

File: src/CLICommand/cli_command.ts [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+const program: CliApiObject = new Command() as CliApiObject;
+
+// Set up command line option parsing
+//const validActions = ['create-drive', 'rename-drive', 'upload-file'];
+program.option('-h, --help', 'Get help');
+//program.option('create-drive', 'action to create a new drive (and its corresponding root folder)');
+program.addHelpCommand(false);
+
+function setCommanderCommand(commandDescriptor: CommandDescriptor, program: CliApiObject): void {
```
Comment (fedellen) [inline src/CLICommand/cli_command.ts:20]: Maybe a small doc block and a small comment above the `const paramType` declaration here?  Its a little hard to follow the what and why here without digging around

**Review by fedellen — APPROVED** [review]
Left a couple minor comments again -- but its looking clean and useful, approved!

File: src/CLICommand/common_context.ts (hunk 8 above)
Comment (matibat) [reply src/CLICommand/common_context.ts]: Check 906681c

File: src/CLICommand/cli_command.test.ts (hunk 9 above)
Comment (matibat) [reply src/CLICommand/cli_command.test.ts]: Removed unused comment 906681c

File: src/CLICommand/cli_command.ts (hunk 10 above)
Comment (matibat) [reply src/CLICommand/cli_command.ts:26]: Adds JSDoc comment b0da89a

**Review by fedellen — APPROVED** [review]
